vortex-layout/src/layouts/chunked/evaluator.rs

```diff
@@ -1,182 +1,110 @@
-                    .row_mask
-                    .slice(chunk_range.start, chunk_range.end)?
-                    .shift(chunk_range.start)?;
-                let chunk_evaluator =
-                    chunk_reader.create_evaluator(chunk_mask, self.expr.clone())?;
-                chunks.push(ChunkState::Pending(chunk_evaluator));
-            }
+        let mut row_offset = 0;
+        for chunk_idx in 0..self.nchunks() {
+            let chunk_reader = self.child(chunk_idx)?;
 
-            self.chunk_states = Some(chunks);
-        }
+            // Figure out the row range of the chunk
+            let chunk_len = chunk_reader.layout().row_count();
+            let chunk_range = row_offset..row_offset + chunk_len;
+            row_offset += chunk_len;
+
+            // Try to skip the chunk based on the row-mask
+            if row_mask.is_disjoint(chunk_range.clone()) {
+                continue;
+            }
 
-        let chunk_states = self
-            .chunk_states
-            .as_mut()
-            .vortex_expect("chunk state not set");
-
-        // Now we try to read the chunks.
-        let mut needed = vec![];
-        for chunk_state in chunk_states.iter_mut() {
-            match chunk_state {
-                ChunkState::Pending(scanner) => match scanner.poll(segments)? {
-                    Poll::Some(array) => {
-                        // Resolve the chunk
-                        *chunk_state = ChunkState::Resolved(Some(array));
-                    }
-                    Poll::NeedMore(segment_ids) => {
-                        // Request more segments
-                        needed.extend(segment_ids);
-                    }
-                },
-                ChunkState::Resolved(_) => {
-                    // Already resolved
+            // If the pruning mask tells us the chunk is pruned (i.e. the expr is ALL false),
+            // then we can just return a constant array.
+            if let Some(pruning_mask) = &pruning_mask {
+                if pruning_mask.value(chunk_idx) {
+                    let false_array = ConstantArray::new(
+                        Scalar::bool(false, dtype.nullability()),
+                        row_mask.true_count(),
+                    );
+                    chunks.push(ready(Ok(false_array.into_array())).boxed_local());
+                    continue;
                 }
             }
-        }
 
-        // If we need more segments, then request them.
-        if !needed.is_empty() {
-            return Ok(Poll::NeedMore(needed));
+            // Otherwise, we need to read it. So we set up a mask for the chunk range.
+            let chunk_mask = row_mask
+                .slice(chunk_range.start, chunk_range.end)?
+                .shift(chunk_range.start)?;
+
+            let chunk_reader = chunk_reader.clone();
+            let expr = expr.clone();
+            chunks.push(async move { chunk_reader.evaluate(chunk_mask, expr).await }.boxed_local());
```

Comment:
This line caught my eye. Ensuring I understood it taught me a lot about async Rust. Thanks!

If we redefined `child` to return a reference, a la:
```rust
    pub(crate) fn child(&self, idx: usize) -> VortexResult<&Arc<dyn LayoutReader>> {
```
then we know the `chunk_reader` lives to the end of this method. If we know that, then we can dispense with the `chunk_reader.clone()`, the `async move` and the `boxed_local`. `chunks` will be `Vec<Pin<Box<dyn Future<_>>>>` which we can join on.

Is there an argument against returning a reference in `child`?